feat(gossipsub): add detailed peer score parameter tracking

lthibault opened the pull request:
## Description

Addresses https://github.com/libp2p/rust-libp2p/issues/6058

Add `PeerScoreParameters` struct and `Behaviour::peer_score_params()` to expose
a granular breakdown of the peer score function defined in the
[gossipsub v1.1 spec](https://github.com/libp2p/specs/blob/master/pubsub/gossipsub/gossipsub-v1.1.md#the-score-function):

- `p1` – time in mesh (summed across topics, pre-weighted)
- `p2` – first message deliveries
- `p3` – mesh message delivery deficit penalty
- `p3b` – mesh failure penalty
- `p4` – invalid message delivery penalty
- `p5` – application-specific score
- `p6` – IP collocation factor penalty
- `p7` – behavioural pattern penalty
- `slow_peer_penalty` – slow peer weight (not in spec, but applied)
- `final_score` – aggregate after topic score cap; equals `peer_score()`

This enables external tooling and operators to inspect which scoring
components are driving a peer's score, rather than only seeing the total.

## Change checklist

- [x] I have performed a self-review of my own code
- [x] I have made corresponding changes to the documentation
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] A changelog entry has been made in the appropriate crates

lthibault committed:
feat(gossipsub): add detailed peer score parameter tracking

Add PeerScoreParameters struct to expose granular breakdown of peer scores:
- Individual contributions for P1-P7 score parameters
- Slow peer penalty tracking
- Final score field

Add peer_score_params() public API method to retrieve detailed score breakdown.

This enables external tooling and debugging to inspect individual components
of the gossipsub peer scoring algorithm per the v1.1 spec.

lthibault committed:
Clean up comment. No functional change.

lthibault committed:
chore(gossipsub): add docs, tests, and changelog for peer_score_params

- Add doc comments to PeerScoreParameters fields and peer_score_params()
- Add Debug derive to PeerScoreParameters
- Add three tests: disabled scoring, final_score consistency, p7 penalty
- Add changelog entry under 0.50.0

jxs commented on protocols/gossipsub/src/behaviour.rs:
```diff
@@ -512,6 +515,19 @@ where
 
+    /// Returns a breakdown of the individual score components for a peer.
+    ///
+    /// Returns `None` if peer scoring is disabled or the peer is unknown.
```
this is not true, if the PeerId doesn't exist in `peer_stats` it returns Some(PeerScoreParameters::default()).
This is also incorrect on `peer_score(&self, peer_id: &PeerId) -> Option<f64>` which should be updated to return None if the peer doesn't exist.

lthibault replied:
Whoops! Fixed.  Thanks for the review!

jxs commented on protocols/gossipsub/src/peer_score.rs:
```diff
@@ -76,10 +76,47 @@ impl PeerScoreState {
+pub struct PeerScoreParameters {
+    /// The final aggregate score. Equal to the sum of all other fields after
+    /// the topic score cap is applied.
+    pub final_score: f64,
```
```suggestion
    pub score: f64,
```

jxs reviewed (commented):
Hi, and thanks for this. Overall looks good, left some comments

lthibault committed:
Update protocols/gossipsub/src/peer_score.rs

Co-authored-by: João Oliveira <hello@jxs.pt>

lthibault committed:
fix(gossipsub): return None for unknown peers in peer_score and peer_score_params

Address review feedback: score_report now returns Option<PeerScoreReport>,
so peer_score() and peer_score_params() correctly return None when the
peer is not in peer_stats, instead of Some(default). Also completes the
final_score -> score rename across doc comments and tests.

jxs committed:
Merge branch 'master' into feature/gossipsub-peer-score-params

jxs commented on protocols/gossipsub/src/behaviour/tests/scoring.rs:
```diff
@@ -50,6 +50,7 @@ use crate::{
     IdentTopic as Topic,
 };
 
+
```
```suggestion
```

jxs commented on protocols/gossipsub/src/behaviour/tests/scoring.rs:
```diff
@@ -2102,5 +2103,93 @@ fn test_subscribe_and_graft_with_negative_score() {
+    assert!(gs1.as_peer_score_mut().score_report(&p2).unwrap().score >= original_score);
+}
+
+// Tests for PeerScoreParameters / peer_score_params().
+
+/// `peer_score_params` returns `None` when scoring is disabled.
```
nit: can we improve the wording on these tests? Like "verifies that.."
